### DISCUSSION TITLE
Forking Renovate automerge

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#9866] [Suggest an Idea]
@rsimha moving the discussion from https://github.com/renovatebot/renovate-approve-bot/issues/55 to here

https://github.com/ampproject/amphtml/pull/34219 is the PR we're hoping to automerge

It's configured for automerge:

![image](https://user-images.githubusercontent.com/6311784/117118172-c19b2c00-ad90-11eb-9e41-09d8f9a2b7e6.png)

It's got an approval:

![image](https://user-images.githubusercontent.com/6311784/117118255-daa3dd00-ad90-11eb-8985-c2de58e80846.png)

There's no *failing* check runs, but when I view the PR when logged in as @renovate-bot it says Merging is blocked:

![image](https://user-images.githubusercontent.com/6311784/117118414-058e3100-ad91-11eb-94e4-e2157dc7dd54.png)

And that's also what we're getting from the GitHub API:

```
{"level":20,"branch":"renovate/core-devdependencies","msg":"PR is configured for automerge","time":"2021-05-05T08:38:40.506Z"}
{"level":20,"branch":"renovate/core-devdependencies","canMergeReason":"mergeStateStatus = BLOCKED","msg":"PR is not ready for merge","time":"2021-05-05T08:38:40.506Z"}
```

**Comment by rsimha** [comment]
This is useful information. Let me investigate at our end. Thank you @rarkins!

**Reply by rsimha** [reply]
This helped me identify a blocking branch protection rule we had in place. I've updated it to allow merging by @renovate-bot.

I'll manually merge this PR and pay attention to the next new update PR and report here with results. Thanks for helping me figure this out!

**Reply by rarkins** [reply]
I can confirm that the bot account got a green Squash and Merge button for that PR. I think the bot would have tried to automerge it if you'd left it open, although possibly it would have tried to merge using the forking-renovate[bot] credentials instead of renovate-bot. Let's see the next time.

**Reply by rsimha** [reply]
We have a new candidate: https://github.com/ampproject/amphtml/pull/34231. PR  is configured for automerge, all commit statuses are green, review approvals have been granted by `renovate-approve`, and I believe @renovate-bot should have the permissions to merge (although IIRC, automerge had not been enabled at the time of PR creation). Will leave this untouched and observe for a while.

**Reply by rarkins** [reply]
OK, don't merge that one. It can automerge if (a) it's all green, and (b) it's up to date with master. But first it looks like it will need a small code change, so probably sometime tomorrow.

**Reply by rarkins** [reply]
@rsimha it has a failing check:

![image](https://user-images.githubusercontent.com/6311784/117420115-8da14180-af1d-11eb-8f8b-355a4e02a583.png)

**Reply by rsimha** [reply]
I've approved that check, they're all green now. I'll continue to keep an eye on the PR and make sure it stays that way. (The bot appears to have force pushed new commits several times, most recently overnight. I assume this has to do with syncing with the `main` branch.)

**Reply by rarkins** [reply]
Yeah, the bot wants to keep the branch up-to-date before automerging

**Reply by rarkins** [reply]
It needed to bring the branch up to date again. I'll try to watch for when it turns green again and run another job manually so that it's hopefully free to automerge.

**Reply by rarkins** [reply]
`"branch":"renovate/build-system-devdependencies","canMergeReason":"mergeStateStatus = BLOCKED","msg":"PR is not ready for merge"`

I think the reason is because the forking-renovate[bot] account is the one which fetches the PR details from GitHub's API, and *it* is blocked. But because of that renovate-bot never tries because it's designed to not attempt to merge PRs which GitHub says are blocked.

I'll need to think more about how to do this. We can't query too much of the API using renovate-bot because it's a regular user and gets only 5000 requests per hour even if it's added to thousands of repos.

**Reply by rsimha** [reply]
Belated comment to say thanks for looking into this and for considering this feature request. I did some homework to see if AMP could switch to the main `renovate` app. Unfortunately, the reasons that caused us to choose `forking-renovate` still remain today.

**Another thought that came to mind:** Now that GitHub has made its [native auto-merge feature](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request#enabling-auto-merge) available to all, is it a viable alternative to have the app simply enable auto-merge at the time of PR creation if `.renovaterc.json` has `automerge` enabled? (The way this would work is that if all checks pass and no merge conflicts arise while they were running, the PR will be auto-merged by GitHub with no further action from the renovate app.) I recognize this is different from the app's current behavior of first checking to make sure the branch is up-to-date and only then merging, but figured it's worth asking, since this could save unnecessary GitHub API requests from the app and eliminate superfluous CI builds.
